src/main/java/net/floodlightcontroller/forwarding/Forwarding.java

```diff
@@ -228,46 +200,661 @@ private void removeExpiredFlowSetId(U64 flowSetId, NodePortTuple avoid, Iterator
-            case NONE:
-                // don't do anything
-                return Command.CONTINUE;
-            case FORWARD_OR_FLOOD:
-            case FORWARD:
-                doForwardFlow(sw, pi, decision, cntx, false);
-                return Command.CONTINUE;
-            case MULTICAST:
-                // treat as broadcast
-                doFlood(sw, pi, decision, cntx);
-                return Command.CONTINUE;
-            case DROP:
-                doDropFlow(sw, pi, decision, cntx);
-                return Command.CONTINUE;
-            default:
-                log.error("Unexpected decision made for this packet-in={}", pi, decision.getRoutingAction());
-                return Command.CONTINUE;
-            }
-        } else { // No routing decision was found. Forward to destination or flood if bcast or mcast.
-            if (log.isTraceEnabled()) {
-                log.trace("No decision was made for PacketIn={}, forwarding", pi);
-            }
+            switch(decision.getRoutingAction()) {
+                case NONE:
+                    // don't do anything
+                    return Command.CONTINUE;
+
+                case FORWARD_OR_FLOOD:
+                case FORWARD:
+                    doL2ForwardFlow(sw, pi, decision, cntx, false);
+                    return Command.CONTINUE;
+
+                case MULTICAST:
+                    // treat as broadcast
+                    doFlood(sw, pi, decision, cntx);
+                    return Command.CONTINUE;
+
+                case DROP:
+                    doDropFlow(sw, pi, decision, cntx);
+                    return Command.CONTINUE;
+
+                default:
+                    log.error("Unexpected decision made for this packet-in={}", pi, decision.getRoutingAction());
+                    return Command.CONTINUE;
+
+            }
+        }
+        else { // No routing decision was found
+            switch(determineRoutingType()) {
+                case FORWARDING:
+                    // L2 Forward to destination or flood if bcast or mcast
+                    if (log.isTraceEnabled()) {
+                        log.trace("No decision was made for PacketIn={}, do L2 forwarding", pi);
+                    }
+                    doL2Forwarding(eth, sw, pi, decision, cntx);
+                    break;
+
+                case ROUTING:
+                    // TODO: IPv6 not consider for now
+                    if (log.isTraceEnabled()) {
+                        log.trace("No decision was made for PacketIn={}, do L3 routing", pi);
+                    }
+
+                    if (!getGatewayInstance(npt).isPresent() && !getGatewayInstance(sw.getId()).isPresent()) {
+                        log.info("Could not locate virtual gateway instance for DPID {}, port {}", sw.getId(), inPort);
+                        break;
+                    }
+
+                    if (getGatewayInstance(sw.getId()).isPresent()) {
```

Comment:
Same here WRT use of Optionals. Get the Optional first and save it to a local variable. Then, work from that variable. (Note this advice isn't limited to just Optionals. It goes for any getter that might return an unusable (i.e. null, perhaps) value.)

Reply:
Fixed.